### DISCUSSION TITLE
Romm support

### DISCUSSION THREAD

**Discussion by binarygeek119** [retronas/discussions#11] [Ideas]
https://github.com/zurdi15/romm

RomM (Rom Manager) is a web based retro roms manager integrated with IGDB.

should also add MariaDB and redis as it works better then.

**Comment by zacaronii** [comment]
I’m also interested in whether or not this is something retronas would be able to integrate with in any capacity. I am using RomM mostly as a pretty way to view my game library. 

However, they also support playing roms through emulatorJS, as well as saves and states. They have mentioned “device sync”, as a feature they are very interested in implementing but they do not have this yet. 

I think the ability to use this with retronas would be awesome. I might even look at fiddling around with this, myself.

Also, as this is my first post on any part of the retronas project, I want to take the opportunity to give kudos to those that have put work into the project. My knowledge in Linux systems is all pretty surface level, but I’ll see if I can learn enough to try to contribute to the project, as well!

**Reply by amymowatt** [reply]
this is now in the stable builds. there are even api tokens for third-party clients. would really love to see tight romm integration with save support for each user.

would also love to see support for serving bios files to romm along with sub-folders for each game for hacks, translations, mods etc.

**Comment by drinksomewhiskey** [comment]
@Sairuk I feel like this can be done similar to the analogue pocket support. Romm just needs symlinks organized by its own folder structure.

**Reply by sairuk** [reply]
@minorOffense is looking at this atm, there are some caveats with Romm that symlinking wont work at the folder level

First run at this support was here https://github.com/retronas/retronas/pull/457

**Comment by minorOffense** [comment]
I have a prototype working. I'm just currently fighting with Ansible. I'm just trying to figure out how to get it to download the script to flatten the rom directory and run it.

Then I think it'll be good to review.

**Reply by minorOffense** [reply]
https://github.com/minorOffense/retronas/pull/2/files

Oh and then the fun task of updating the entire systems array to map to romm's naming scheme :-/

**Reply by sairuk** [reply]
i've added some comments to that pull request this morning

**Reply by sairuk** [reply]
where are you getting the structure information from, is there a source of truth for all directories we can pull down and process programmatically?

**Reply by minorOffense** [reply]
I was going by their mapping in the examples directory: https://github.com/rommapp/romm/blob/master/examples/config.batocera-retrobat.yml#L16

Comparing to how batocera was setup between retronas and romm. I think their list is based on the igdb database. So I was going to start there and see how it maps or not. I assumed I would have to ask the romm folks about that.

**Reply by drinksomewhiskey** [reply]
This could be helpful: https://docs.romm.app/4.0.0/Getting-Started/Folder-Structure/

**Comment by drinksomewhiskey** [comment]
Hi @sairuk and @minorOffense - hope all is well. 

Any updates on this by chance?

**Reply by minorOffense** [reply]
Sorry I caught the flue and then pneumonia before xmas and then haven't gotten back to this since. I'll try to make time this weekend.

**Reply by sairuk** [reply]
ouch hope you are doing OK,  i'm happy to pick this up from the retronas side i.e. https://github.com/retronas/retronas/issues/464 if you aren't able to run with it. I was planning to if I hadn't heard from you by start of feb.

**Reply by drinksomewhiskey** [reply]
Hope all is well!

**Comment by amymowatt** [comment]
with the romm cifs script added, are files renamed or uploaded from romm reflected in the rest of retronas dirs?

**Reply by sairuk** [reply]
if you use the romm share the files will not be linked back to the main retronas structure; the files are scanned and hard linked from the existing retronas directories to the romm folder structure. We should probably make the romm share read only by default unless romm itself requires it to be writable, i'd need to look into that.

to work with romm, users should upload files through our existing retronas (share) folder structure then the scanner will hard link them back to the romm folder structure

**Reply by sairuk** [reply]
this was a good question, thanks, i can add my response in part to some doco that needs to be written.

**Reply by sairuk** [reply]
as a follow up in the latests commits the main romm share is read only and users should upload files through the retronas share then run the appropriate script to populate the romm structure.

**Comment by amymowatt** [comment]
is there a particular way we should be organizing files and folders in the generic rom folder to make dlc, updates, hacks, mods etc appear as a sub folder for the relevant game in romm when accessing the romm cifs share?

**Reply by sairuk** [reply]
generic folders are a retronas structure thing and i don't have any plans to move to an alternate structure for a single project. if users want the extra folders romm supports they would go into the main romm folder meaning they would only be available to romm. 

If the romm project has info on configuring its structure against other projects (i.e. batocera) there should be no reason users can't apply their config with the retronas romm share (keeping in mind it is read only now)

for everyones reference we are referring to this structure
```
        │  ├─ system/
        │  │  ├─ game_3.imarom
        │  │  └─ game_4/
        │  │     ├─ game_4.imaromalso
        │  │     ├─ dlc
        │  │     │  ├─ game_4_dlc_1.7z
        │  │     │  └─ game_4_dlc_2.7z
        │  │     ├─ hack
        │  │     │  └─ game_4_hardmode.rar
        │  │     ├─ manual
        │  │     │  └─ game_4_manual.pdf
        │  │     ├─ mod
        │  │     │  └─ game_4_crazy_mode.zip
        │  │     ├─ patch
        │  │     │  └─ game_4_patch_v1.1.zip
        │  │     ├─ update
        │  │     ├─ demo
        │  │     ├─ translation
        │  │     └─ prototype
```

there is no automated way (again atm) to make these directories in the romm folder, however we could probably add a script you could run to generate things (perhaps with manual clean up of old software)

you need to keep in mind that the software files are scripted and hardlinked to this directory structure and moving them would break the current implementation without some modification of the @minorOffense script handling that.
